### PR TITLE
implements search on null/notnull metadata

### PR DESCRIPTION
**Pull request by ArtificialOwl**
Add the possiblity to search on null/not null value on metadata.
This is a test-feature for the Photos App.

This PR requires to add this line
```
      '{DAV:}is-defined'    => Operator::class,
```
in https://github.com/nextcloud/3rdparty/blob/master/icewind/searchdav/src/DAV/QueryParser.php#L78

```
[...]
     <d:where>
        <d:and>
          <d:not>
          <d:is-defined>
            <d:prop><nc:metadata-test/></d:prop>
            </d:is-defined>
          </d:not>
        </d:and>
[...]
```

Need:
- [x] https://github.com/icewind1991/SearchDAV/pull/13
- [x] https://github.com/nextcloud/3rdparty/pull/1643

### CODE REVIEW

File: lib/private/Files/Cache/SearchBuilder.php [hunk 1]
```diff
@@ -57,6 +58,7 @@ class SearchBuilder {
 		ISearchComparison::COMPARE_GREATER_THAN_EQUAL => 'lt',
 		ISearchComparison::COMPARE_LESS_THAN => 'gte',
 		ISearchComparison::COMPARE_LESS_THAN_EQUAL => 'gt',
+		ISearchComparison::COMPARE_NULL => 'isNotNull',
```
Comment (artonge) [inline lib/private/Files/Cache/SearchBuilder.php:61]:
```suggestion
		ISearchComparison::COMPARE_NOT_NULL => 'isNotNull',
```

File: apps/dav/lib/Files/FileSearchBackend.php [hunk 2]
```diff
@@ -382,7 +382,9 @@ private function transformSearchOperation(Operator $operator) {
 					return new SearchComparison($trimmedType, $this->mapPropertyNameToColumn($property), $value);
 				}
 
-				// no break
+			case Operator::OPERATION_IS_DEFINED:
+				$property = $operator->arguments[0];
+				return new SearchComparison('null', substr($property->name, strlen(FilesPlugin::FILE_METADATA_PREFIX)), '', IMetadataQuery::EXTRA);
```
Comment (artonge) [inline apps/dav/lib/Files/FileSearchBackend.php:387]: Maybe it should not be limited to the metadata?

**Review by artonge — COMMENTED** [review]
Nice

File: apps/dav/lib/Files/FileSearchBackend.php (hunk 2 above)
Comment (ArtificialOwl) [reply apps/dav/lib/Files/FileSearchBackend.php:387]: let's check if it is working with metadata first !

File: lib/private/Files/Cache/SearchBuilder.php (hunk 1 above)
Comment (ArtificialOwl) [reply lib/private/Files/Cache/SearchBuilder.php:61]: check the name of the array :]

**Comment by icewind1991** [comment]
might be cleaner to just transform the dav operator to `field == null` `!(field == null)` instead of adding a new search operator

**Comment by artonge** [comment]
> might be cleaner to just transform the dav operator to `field == null` `!(field == null)` instead of adding a new search operator

Do you mean in the DAV request ? The issue is that we are casting the type of the property here: https://github.com/nextcloud/server/blob/b79a21d29d575fe42ae74aa21bc1a7d08546367c/apps/dav/lib/Files/FileSearchBackend.php#L418-L436
And as the property is used in another comparison in the request, it is assumed to be an integer. So using the appropriate operator would be cleaner I think.

**Comment by artonge** [comment]
Pushed 3rdparty changes

**Review by ChristophWurst — APPROVED** [review]
:+1: for the caldav changes

**Review by AndyScherzinger — APPROVED** [review]

**Comment by ChristophWurst** [comment]
adds new public API -> needs docs

**Pull request merged**